### PR TITLE
Add support for temporal types

### PR DESCRIPTION
**Pull request by gitbuda**

### CODE REVIEW

**Comment by gitbuda** [comment]
@kostasrim here we have a little bit bigger problem because there is no obvious mapping between `mg_local_time` and `mg_duration`. Options are:

1. Create an arbitrary `Napi::Object` (what's currently showcased in the code) -> Less work, the idea would be just to preserve info, without spending much time on the interface of these objects (just passing what `mgclient` sent).
2. Create an arbitrary `Napi::ObjectWrap`, in other words, create a custom class in C++ that will encode all `mgclient` stuff (a special type) -> More work, essentially implementing https://tc39.es/proposal-temporal/docs/ (if we go this route, I would spend more time on actually implementing the right API).

Any thoughts here are welcome!

**Comment by gitbuda** [comment]
The decision at this point is to proceed with option 1.

**Comment by gitbuda** [comment]
@kostasrim tests are missing (I'll add that once the Memgraph Docker becomes available, refactor is required to make tests better, that will come later), but the implementation of option 1. should be done, feel free to make the review of that part :smile:

File: index.js [hunk 1]
```diff
@@ -15,6 +15,62 @@
...
+function createMgLocalTime(nanoseconds) {
+  return {
+    "objectType": "local_time",
+    "nanoseconds": nanoseconds,
+  }
+}
+
+/**
+  * Create Memgraph compatible local date time object.
+  * @param {number} seconds - The number of seconds since 1970-01-01 (unix epoch).
```
Comment (kostasrim) [inline index.js:46]: and the local_time!

File: README.md [hunk 2]
```diff
@@ -44,3 +44,22 @@ To test ([Docker](https://docs.docker.com/engine/install) is required) run:
...
+
+Suitable JS type to store Memgrpah temporal types don't exist. In particular,
+it's impossible to convert `mg_duration` and `mg_local_time` to the `Date`
+type. Since [the temporal
+specification](https://github.com/tc39/proposal-temporal) is not yet widely
+supported, the decision was to expose plain JS objects (dict) with the exact
+fields `mgclient` is providing. In addition, when possible (`mg_date` and
+`mg_local_date_time`), are converted into objects which have `date` property,
+which in fact, is the `Date` representation of these types. Keep in mind the
+loss of precision because `Date` can only store millisecond precision, but the
```
Comment (kostasrim) [inline README.md:60]: This is slightly misleading, because `Date` is actually `Datetime` in JS. Maybe reword this to:

JS `Date` time fields can only store up to `milliseconds` precision. However,  Memgraph supports microsecond precision for the local time and therefore any use of this class can potentially cause loss of information.

File: index.d.ts [hunk 3]
```diff
@@ -29,6 +29,47 @@ export namespace Memgraph {
...
+  *
+  * @param {number} nanoseconds - The number of nanoseconds since midnight.
+  */
+export function createMgLocalTime(nanoseconds: number): {
+    objectType: string;
+    nanoseconds: number;
+};
+/**
+  * Create Memgraph compatible local date time object.
+  * @param {number} seconds - The number of seconds since 1970-01-01 (unix epoch).
```
Comment (kostasrim) [inline index.d.ts:51]: maybe add the `h:m:s` as well ?

Comment (kostasrim) [inline index.d.ts:51]:
```suggestion
  * @param {number} seconds - The number of seconds since 1970-01-01T00:00:00 (unix epoch).
```

File: index.js [hunk 4]
```diff
@@ -15,6 +15,62 @@
...
+  return {
+    "objectType": "local_date_time",
+    "seconds": seconds,
+    "nanoseconds": nanoseconds,
+  }
+}
+
+/**
+  * Create Memgraph compatible duration object.
+  * NOTE: Semantically Memgraph duration is a sum of all
```
Comment (kostasrim) [inline index.js:59]: That's an implementation detail! Maybe remove the comments (we store everything in microseconds) but I don't see how that information is useful to the clients? Am I missing something?

File: src/glue.cpp [hunk 5]
```diff
@@ -20,16 +20,62 @@ namespace nodemg {
 
 Napi::Value MgStringToNapiString(Napi::Env env, const mg_string *input_string) {
   Napi::EscapableHandleScope scope(env);
-
```
Comment (kostasrim) [inline src/glue.cpp:23]: Dead removed line? Is this intentional?

File: src/glue.cpp [hunk 6]
```diff
@@ -258,6 +318,50 @@ std::optional<Napi::Value> MgValueToNapiValue(Napi::Env env,
   }
 }
 
+std::optional<int64_t> GetInt64Value(Napi::Object input,
+                                     const std::string &key) {
+  if (!input.Has(key)) return std::nullopt;
+  auto key_value = input.Get(key);
+  if (!key_value.IsBigInt()) return std::nullopt;
+  bool lossless;
```
Comment (kostasrim) [inline src/glue.cpp:326]:
```suggestion
  bool lossless{false};
```

File: src/glue.cpp [hunk 7]
```diff
@@ -316,8 +420,52 @@ std::optional<mg_value *> NapiValueToMgValue(Napi::Env env,
       return std::nullopt;
     }
     output_value = mg_value_make_list(*maybe_mg_list);
+    // NOTE: The "dispatch" below could be implemented in a much better way
+    // (probably the whole function as well), but that's a nice exercise for the
+    // future.
```
Comment (kostasrim) [inline src/glue.cpp:425]: :eyes:

**Comment by kostasrim** [comment]
Minor comments, good work!

File: index.js (hunk 4 above)
Comment (gitbuda) [reply index.js:59]: I think it's not just an implementation detail; the question is, what's the meaning of these three fields? Now I see it's also wrong here... I'll update + we can discuss :smile:

File: index.d.ts (hunk 3 above)
Comment (gitbuda) [reply index.d.ts:51]: This is auto-generated, I'll fix it in the `index.js`

File: src/glue.cpp (hunk 5 above)
Comment (gitbuda) [reply src/glue.cpp:23]: Yes, because I thought these empty lines made sense in the first place, but I don't think so anymore...

File: index.d.ts (hunk 3 above)
Comment (gitbuda) [reply index.d.ts:51]: I'll do this manually because it has to be changed in the `index.js` file.

File: index.js (hunk 1 above)
Comment (gitbuda) [reply index.js:46]: What do you exactly mean here? This: `@param {number} seconds - The number of seconds since 1970-01-01T00:00:00 (Unix epoch) and the local_time.`? If so, is that correct actually because, all time is expressed with the 0 offset?

File: README.md (hunk 2 above)
Comment (gitbuda) [reply README.md]: Better now?

File: index.js (hunk 4 above)
Comment (gitbuda) [reply index.js:60]: I got a bit confused here... Since we don't have an equivalent concept in the Node.js case, the question is, are these two:
```
pub(crate) fn mg_value_duration(mg_value: *const bindings::mg_value) -> Duration {
    let c_duration = unsafe { bindings::mg_value_duration(mg_value) };
    let days = unsafe { bindings::mg_duration_days(c_duration) };
    let seconds = unsafe { bindings::mg_duration_seconds(c_duration) };
    let nanoseconds = unsafe { bindings::mg_duration_nanoseconds(c_duration) };
    Duration::days(days) + Duration::seconds(seconds) + Duration::nanoseconds(nanoseconds)
}

pub(crate) fn duration_to_mg_duration(input: &Duration) -> *mut bindings::mg_duration {
    let nsec_in_sec = 1_000_000_000_i64;
    let mut nanoseconds = input.num_nanoseconds().unwrap();
    let days = input.num_days();
    nanoseconds -= days * 24 * 60 * 60 * nsec_in_sec;
    let seconds = input.num_seconds() - days * 24 * 60 * 60;
    nanoseconds -= seconds * nsec_in_sec;
    unsafe { bindings::mg_duration_make(0, days, seconds, nanoseconds) }
}
```
(I'm borrowing the Rust code here), valid implementations. If yes, I think a user has to be aware of that, it doesn't matter how it's implemented, but it matters what the `nanoseconds` field is about, or what's the relation between `seconds` and `nanoseconds`.

File: index.js (hunk 1 above)
Comment (kostasrim) [reply index.js]: I mean that if you have the date `1970-01-01T00:00:01` this is `0 days` since the unix epoch + `00:00:01` (1 second since the unix epoch). Hence, you will receiver 0 (days in seconds) + 1 (second the time point) = 1 second since the unix epoch

File: index.js (hunk 4 above)
Comment (kostasrim) [reply index.js:60]: If Duration is rust Duration, that is, https://doc.rust-lang.org/stable/std/time/struct.Duration.html then yes the implementation is correct. There might be a slight mistake, I think `input.num_nanoseconds() should be input.as_nanos()`.  

For the rest, this is a serialization format of packstream and IS8601 which are an implementation detail. Moreover, we slightly diverge from that (we don't fully support ISO 8601 for Duration -- only in a way for the neo4j driver to work with memgraph)

Comment (gitbuda) [reply index.js:60]: I understand that, but it's not about serialization, it's about what to do with these objects, how does a user know what the object is? With Rust is easy because there is documentation of the type (this https://docs.rs/chrono/0.4.19/chrono/struct.Duration.html), but in the Node.js case, we have a custom/new object which nobody understands. That's what I had/have in mind :smile:

File: index.js (hunk 1 above)
Comment (gitbuda) [reply index.js]: Aha, if I understand correctly, this is now an opposite discussion to the one about the duration :smile: Isn't this the implementation detail? A user doesn't care about how is that serialized, the user has the number of seconds, and that has to be explained?
